torchtitan/models/attention.py

```diff
@@ -15,244 +15,213 @@
+
+
+def get_fixed_block_mask_mod(
+    mask_mod: _mask_mod_signature, fixed_block_size: int
+) -> _mask_mod_signature:
+    """
+    Given an arbitrary mask_mod, divide the input sequence to blocks
+    and only allow attention within the same block.
+
+    Args:
+        mask_mod: The mask mod to apply to the documents
+        fixed_block_size: The number of tokens in each block.
+    """
+
+    # Credit to @drisspg.
+    def blocked_mask_mod(
+        b: torch.Tensor, h: torch.Tensor, q_idx: torch.Tensor, kv_idx: torch.Tensor
+    ):
+        # Get the block index of the query and key
+        q_block = q_idx // fixed_block_size
+        kv_block = kv_idx // fixed_block_size
+        # Only allow attention within the same block
+        same_block = q_block == kv_block
+        # Apply the original mask mod
+        inner_mask = mask_mod(b, h, q_idx % fixed_block_size, kv_idx % fixed_block_size)
+
+        return same_block & inner_mask
+
+    blocked_mask_mod.__name__ = (
+        f"blocked_mask_mod_{mask_mod.__name__}_fixed_block_size_{fixed_block_size}"
+    )
+
+    return blocked_mask_mod
+
+
+_compiled_create_block_mask: Callable | None = None
+
+
+def get_create_mask_fn(
+    use_flex_attn: bool,
     cp_mesh: torch.distributed.device_mesh.DeviceMesh | None = None,
-) -> None:
+) -> Callable:
+    """
+    Get a function that creates an attention mask. This is currently only
+    mneaningful when FlexAttention is used. For SDPA, the returned
+    function should never be called as SDPA doesn't support varlen yet.
+
+    Note: the returned create_mask_fn will cache the masks if the
+    input is the same.
+
+    Args:
+        use_flex_attn: Whether to use FlexAttention or not.
+        cp_mesh: The device mesh to use for creating the block mask. If None, the
+            function will return a function that creates a block mask for the current
+            device mesh.
+
+    Returns:
+        create_mask_fn: a function hat creates an attention mask. This function
+        will catch the results if the input is the same.
+    """
 
     # This is not functional yet because we currently gate the use of Flex + CP
     # while we continue debugging accuracy issues. However, we want to evaluate
     # the user experience with CP enabled.
+    global _compiled_create_block_mask
     if cp_mesh is not None:
-        FlexAttention.compiled_create_block_mask = functools.partial(
-            create_cp_block_mask, device_mesh=cp_mesh
+        from torch.distributed.tensor.experimental._attention import _DispatchMode
+
+        torch.distributed.tensor.experimental._attention._dispatch_mode = (
+            _DispatchMode.MODULE_WRAPPER
+        )
+        if _compiled_create_block_mask is None:
+            _compiled_create_block_mask = functools.partial(
+                create_cp_block_mask, device_mesh=cp_mesh
+            )
+    elif _compiled_create_block_mask is None:
+        _compiled_create_block_mask = torch.compile(create_block_mask)
```

Comment:
do we need this? is this somehow a lazy call?

Reply:
I'll remove create_cp_block_mask, the new UX design won't require create_cp_block_mask anymore.